Field Access Patterns for ChainTipSubscriber, MempoolKey, MempoolValue, BestTip, StatusError, AtomicStatus

ala-mode opened the pull request:
Improves #516

ala-mode committed:
named field in AtomicStatus

ala-mode committed:
GetDifficultyResponse named field

nachog00 commented on zaino-state/src/status.rs:
```diff
@@ -20,22 +20,27 @@ use std::{
 ///   TODO: Refine error code spec.
 #[derive(Debug, Clone)]
-pub struct AtomicStatus(Arc<AtomicUsize>);
+pub struct AtomicStatus {
```
having a named field is already an improvement in clarity (compared with obscured ".0"). But is think it would be even better to work with getters ( .counter() ). Cause that way you aren't tied to a specific internal representation.

ala-mode replied:
I agree on this point.

To answer the above comment first, I _was_ planning on doing everything reasonable with one PR, and only break up the PRs if I needed to. I'm trying to make each commit its own pattern-alteration, or at least not have more than one pattern change in any given commit.

If we want to add more or change logic to address these kind of observations, I propose putting _those_ into more issues, and breaking them off one at a time.

I'd like to try to get these patterns changed in the next few days, so it shouldn't go too stale.

nachog00 commented on zaino-state/src/status.rs:
```diff
@@ -20,22 +20,27 @@ use std::{
     /// Loads the value held in the AtomicStatus
     pub fn load(&self) -> usize {
-        self.0.load(Ordering::SeqCst)
+        self.counter.load(Ordering::SeqCst)
```
see, if this was `counter()` then say we did want to move back to tuple internals or simply change it somehow, then while  `.counter` could stop being a valid accessor and one would have to update the usage everywhere (not a massive hit on this simpler case ofc, but the alternative is simple too), a `.counter()` call could be left untouched everywhere and simply update it's definition to get the right value.

ala-mode replied:
Yes, I like this.

Like I said, maybe we could break these kinds of improvements up into additional PRs, and have this PR be in effect 'non-changing'?

nachog00 commented on zaino-state/src/status.rs:
```diff
@@ -20,22 +20,27 @@ use std::{
     }
 
     /// Sets the value held in the AtomicStatus
+    // TODO do we need to have status be usize here? why not strong types?
```
I wonder this too. I would pursue this change.

ala-mode committed:
Mempoolkey named field txid

ala-mode committed:
Merge branch 'dev' into field_access_patterns

ala-mode committed:
add doc comments

ala-mode committed:
update tests, remove dbg!

ala-mode committed:
update mempool, remove dbg!

ala-mode committed:
MempoolValue named field

ala-mode committed:
MempoolValue add tests

ala-mode committed:
refine comment

ala-mode committed:
rename variable in LightWalletIndexer utxo, named field for ChainTipSubscriber

ala-mode committed:
StatusError named field

ala-mode committed:
StatusError named field continued

ala-mode committed:
named variable returns for build_lightclients()

ala-mode committed:
WIP best_tip tuple field to new BestTip struct

ala-mode committed:
WIP BestTip struct

ala-mode committed:
WIP BestTip struct

ala-mode committed:
rename variable txid to tx_hash

ala-mode committed:
Remove redundantly named variable

ala-mode committed:
name tuple vars in z_get_address_utxos

ala-mode committed:
remove comment

ala-mode commented:
I'm opening this for review because I think I attacked the problem significantly enough to make an impact.

Due to @idky137 's remarks, I tried to avoid changing anything where we are using zebra-types directly (particularly `Height` and `BlockHash`), and anything in `local_cache`. I'm not 100% confident I succeeded, but that was the aim and I hope a review will confirm these changes are appropriate and will help. I don't think I changed any logic such that Zaino will run any differently.

The work could continue: above, @nachog00 suggests a few things. One is to more strongly type the `status` - which is a change I support, but I think would need a little logic overhaul. A related idea brought up is to use getter(/setter?) patterns and create a `.counter()` function.

There are probably many more things that could be addressed as well. Looking at the code, these changes are diverging (in a good way, I think) from patterns found in `zebra` - so one thought would be to see if ZF is interested in following our lead on this.

Open to any thoughts.

ala-mode committed:
WIP: errors with getdifficulty

ala-mode committed:
Merge branch 'dev' into field_access_patterns

ala-mode committed:
WIP revert GetDifficulty

ala-mode committed:
finish revert GetDifficulty

nachog00 commented:
I agree that we shouldn't simply copy zebra's approach if it is a sub-optimal pattern. Unless there's explicit reasons to stay coherent with it.

nachog00 committed:
Trigger CI with dummy change in Cargo.toml
- Added a comment in Cargo.toml to trigger CI
- This change will be reverted shortly

nachog00 committed:
Revert "Trigger CI with dummy change in Cargo.toml"

This reverts commit d205d0d662e29b162c8a44392977067e6cd28f0b.
